File: WinForms/VB/VertGridMainDemo/CodeExamples/FocusAndNavigation.vb
Imports System.Windows.Forms
Imports DevExpress.Data.Filtering
Imports DevExpress.DXperience.Demos.CodeDemo
Imports DevExpress.DXperience.Demos.CodeDemo.Data
Imports DevExpress.XtraEditors
Imports DevExpress.XtraVerticalGrid
Imports DevExpress.XtraVerticalGrid.Rows
Imports SampleDataSourceFile = DevExpress.XtraVerticalGrid.Demos.CodeExamples.SampleDataCS
Imports DevExpress.Utils
Imports System.Drawing

Namespace DevExpress.XtraVerticalGrid.Demos.CodeExamples

    <DevExpress.DXperience.Demos.CodeDemo.CodeExampleClass("Focus and navigation", "FocusAndNavigation.cs")>
    Public Module FocusAndNavigation

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleSetUp>
        Public Function SetUp(ByVal sampleHost As DevExpress.XtraEditors.XtraUserControl) As Object()
            Dim vGridControl As DevExpress.XtraVerticalGrid.VGridControl = New DevExpress.XtraVerticalGrid.VGridControl()
            vGridControl.Dock = System.Windows.Forms.DockStyle.Fill
            vGridControl.Parent = sampleHost
            vGridControl.RowHeaderWidth = DevExpress.Utils.ScaleUtils.ScaleValue(120)
            vGridControl.RecordWidth = DevExpress.Utils.ScaleUtils.ScaleValue(140)
            vGridControl.DataSource = DevExpress.DXperience.Demos.CodeDemo.Data.SampleData.GetData(50)
            Return New Object() {vGridControl}
        End Function

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleTearDown>
        Public Sub TearDown(ByVal sampleHost As DevExpress.XtraEditors.XtraUserControl)
            TryCast(sampleHost.Controls(CInt((0))), DevExpress.XtraVerticalGrid.VGridControl).Dispose()
        End Sub

#Region "Focus and Navigation"
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Scroll horizontally by rows"), DevExpress.XtraVerticalGrid.Demos.CodeExamples.SampleDataCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("LeftVisibleRecord")>
        Public Sub ScrollViewVerticallyByRows(ByVal vGridControl As DevExpress.XtraVerticalGrid.VGridControl)
            Dim b As DevExpress.XtraEditors.SimpleButton = New DevExpress.XtraEditors.SimpleButton With {.Parent = vGridControl.Parent, .Text = "Scroll right", .Dock = System.Windows.Forms.DockStyle.Top}
            ' The index of the leftmost visible record
            AddHandler b.Click, Sub(s, e) vGridControl.LeftVisibleRecord += 1
            Dim b1 As DevExpress.XtraEditors.SimpleButton = New DevExpress.XtraEditors.SimpleButton With {.Parent = vGridControl.Parent, .Text = "Scroll left", .Dock = System.Windows.Forms.DockStyle.Top}
            AddHandler b1.Click, Sub(s, e) vGridControl.LeftVisibleRecord -= 1
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Scroll horizontally by pixels"), DevExpress.XtraVerticalGrid.Demos.CodeExamples.SampleDataCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("LeftVisibleRecordPixel")>
        Public Sub ScrollViewVerticallyByPixels(ByVal vGridControl As DevExpress.XtraVerticalGrid.VGridControl)
            ' Toggles pixel-based vertical scrolling on or off
            vGridControl.OptionsBehavior.AllowAnimatedScrolling = True
            Dim b As DevExpress.XtraEditors.SimpleButton = New DevExpress.XtraEditors.SimpleButton With {.Parent = vGridControl.Parent, .Text = "Scroll right", .Dock = System.Windows.Forms.DockStyle.Top}
            ' The horizontal scroll position of the leftmost record.
            AddHandler b.Click, Sub(s, e) vGridControl.LeftVisibleRecordPixel += 6
            Dim b1 As DevExpress.XtraEditors.SimpleButton = New DevExpress.XtraEditors.SimpleButton With {.Parent = vGridControl.Parent, .Text = "Scroll left", .Dock = System.Windows.Forms.DockStyle.Top}
            AddHandler b1.Click, Sub(s, e) vGridControl.LeftVisibleRecordPixel -= 6
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Prohibit focusing a column"), DevExpress.XtraVerticalGrid.Demos.CodeExamples.SampleDataCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("FocusedColumnChanged", "FocusedColumn", "PrevFocusedColumn")>
        Public Sub ProhibitFocusingColumn(ByVal vGridControl As DevExpress.XtraVerticalGrid.VGridControl)
            ' This event fires each time a user selects another column
            AddHandler vGridControl.FocusedRowChanged, Sub(s, e)
                If e.Row.VisibleIndex Mod 2 = 1 Then
                    vGridControl.FocusedRow = e.OldRow
                End If
            End Sub
        End Sub
#End Region
    End Module
End Namespace

File: WinForms/CS/VertGridMainDemo/CodeExamples/FocusAndNavigation.cs
using System.Windows.Forms;
using DevExpress.Data.Filtering;
using DevExpress.DXperience.Demos.CodeDemo;
using DevExpress.DXperience.Demos.CodeDemo.Data;
using DevExpress.XtraEditors;
using DevExpress.XtraVerticalGrid;
using DevExpress.XtraVerticalGrid.Rows;
using SampleDataSourceFile = DevExpress.XtraVerticalGrid.Demos.CodeExamples.SampleDataCS;
using DevExpress.Utils;
using System.Drawing;

namespace DevExpress.XtraVerticalGrid.Demos.CodeExamples {
    [CodeExampleClass("Focus and navigation", "FocusAndNavigation.cs")]
    public static class FocusAndNavigation {
        [CodeExampleSetUp]
        public static object[] SetUp(XtraUserControl sampleHost) {
            VGridControl vGridControl = new VGridControl();
            vGridControl.Dock = DockStyle.Fill;
            vGridControl.Parent = sampleHost;
            vGridControl.RowHeaderWidth = ScaleUtils.ScaleValue(120);
            vGridControl.RecordWidth = ScaleUtils.ScaleValue(140);
            vGridControl.DataSource = SampleData.GetData(50);
            return new object[] { vGridControl };
        }
        [CodeExampleTearDown]
        public static void TearDown(XtraUserControl sampleHost) {
            (sampleHost.Controls[0] as VGridControl).Dispose();
        }

        #region Focus and Navigation

        [CodeExampleCase("Scroll horizontally by rows"), SampleDataSourceFile]
        [CodeExampleUnderlineTokens("LeftVisibleRecord")]
        public static void ScrollViewVerticallyByRows(VGridControl vGridControl) {
            SimpleButton b = new SimpleButton { Parent = vGridControl.Parent, Text = "Scroll right", Dock = DockStyle.Top };
            b.Click += (s, e) => {
                // The index of the leftmost visible record
                vGridControl.LeftVisibleRecord++;
            };

            SimpleButton b1 = new SimpleButton { Parent = vGridControl.Parent, Text = "Scroll left", Dock = DockStyle.Top };
            b1.Click += (s, e) => {
                vGridControl.LeftVisibleRecord--;
            };
        }

        [CodeExampleCase("Scroll horizontally by pixels"), SampleDataSourceFile]
        [CodeExampleUnderlineTokens("LeftVisibleRecordPixel")]
        public static void ScrollViewVerticallyByPixels(VGridControl vGridControl) {
            // Toggles pixel-based vertical scrolling on or off
            vGridControl.OptionsBehavior.AllowAnimatedScrolling = true;

            SimpleButton b = new SimpleButton { Parent = vGridControl.Parent, Text = "Scroll right", Dock = DockStyle.Top };
            b.Click += (s, e) => {
                // The horizontal scroll position of the leftmost record.
                vGridControl.LeftVisibleRecordPixel += 6;
            };

            SimpleButton b1 = new SimpleButton { Parent = vGridControl.Parent, Text = "Scroll left", Dock = DockStyle.Top };
            b1.Click += (s, e) => {
                vGridControl.LeftVisibleRecordPixel -= 6;
            };
        }

        [CodeExampleCase("Prohibit focusing a column"), SampleDataSourceFile]
        [CodeExampleUnderlineTokens("FocusedColumnChanged", "FocusedColumn", "PrevFocusedColumn")]
        public static void ProhibitFocusingColumn(VGridControl vGridControl) {
            // This event fires each time a user selects another column
            vGridControl.FocusedRowChanged += (s, e) => {
                if(e.Row.VisibleIndex % 2 == 1) {
                    vGridControl.FocusedRow = e.OldRow;
                }
            };
        }
        #endregion
    }
}
